build_tools/third_party/swiftshader/build_vk_swiftshader.sh

```diff
@@ -15,51 +15,60 @@
 # limitations under the License.
 set -e
 
-# Call this script **from the project root** to build SwiftShader's Vulkan ICD:
+# Call this script to build SwiftShader's Vulkan ICD:
 #   $ bash build_tools/third_party/swiftshader/build_vk_swiftshader.sh
 #
+# The default installation directory is `${HOME?}/.swiftshader`, but it can be
+# overridden as follows:
+#   $ bash build_tools/third_party/swiftshader/build_vk_swiftshader.sh <swiftshader-install-dir>
+#
 # Note that you will need a working CMake installation for this script to
 # succeed. On Windows, Visual Studio 2019 is recommended.
 #
 # Afterwards, set the VK_ICD_FILENAMES environment variable to the absolute
 # path of the corresponding vk_swiftshader_icd.json manifest file so the Vulkan
-# loader on your system loads it, for example:
-#   Windows
-#   $ set VK_ICD_FILENAMES=C:\dev\iree\build-swiftshader\Windows\vk_swiftshader_icd.json
-#
-#   Linux
-#   $ VK_ICD_FILENAMES=/dev/iree/build-swiftshader/Linux/vk_swiftshader_icd.json
+# loader on your system loads it.
 #
 # See https://vulkan.lunarg.com/doc/view/1.1.70.1/windows/loader_and_layer_interface.html
 # for further details about the Vulkan loader and ICDs.
 
-# Check that we're in the project root so our relative paths work as expected.
-if [[ $(basename "$PWD") != "iree" ]]; then
-    >&2 echo "******************************************************"
-    >&2 echo "* This script should be run from IREE's project root *"
-    >&2 echo "******************************************************"
-    exit 1
-fi
+SWIFTSHADER_COMMIT=6287c18b1d249152563f0cb2d5cb0c6d0eb9e3d6
+SWIFTSHADER_DIR="$(mktemp --directory --tmpdir swiftshader_XXXXXX)"
+SWIFTSHADER_INSTALL_DIR="${1:-${HOME?}/.swiftshader}"
```

Comment:
I think this will install swiftshader under the WSL on windows. Is that going to make things weird? The normal place to do this sort of weirdo install on linux would be `/opt/` I think? I'm not sure there's anything we can do here that will work on both windows and linux, so I think it would probably be best to either detect windows with `$OSTYPE` or `uname` or something or just require it be set

Reply:
`$HOME` isn't normally set: https://superuser.com/questions/607105/is-the-home-environment-variable-normally-set-in-windows I happen to have it set under cmder on my machine to `C:\USERS\$USERNAME` though.

You could use
* `$LOCALAPPDATA` (`C:\Users\$USERNAME\AppData\Local`)
* `$USERPROFILE` (`C:\Users\$USERNAME`), basically `$HOME`

see also https://softwareengineering.stackexchange.com/questions/160097/where-should-i-put-configuration-files

Reply:
Geoffrey: The idea was that it would fail on windows (because `"${HOME?}"` isn't set), so the user would just have to choose where to put it. Any reason to prefer `/opt/` over a user install? Linux sure did choose a non-descriptive directory structure.

Scott: I'll add an OS switch with default variables for each os. I'll default windows to `$USERPROFILE\.swiftshader`? Idk. Swiftshader isn't really an "App", but maybe AppData is the right place? (Also I'm assuming dotfiles mean the same thing on windows that they do in Linux/macOS).

Reply:
`$USERPROFILE\.swiftshader` seems more reasonable to me than `AppData`. I have a bunch of dotfiles in `$USERPROFILE` like `.android`, `.keras`, and `.vscode`. I only have one dotfile under `AppData`.

Personally I'll build SwiftShader either next to projects that use it or in a dedicated directory like `D:\dev\tools`

Reply:
It's not a dotfile though, it's a whole piece of software being installed. That's why I suggested `/opt/`

> The idea was that it would fail on windows (because "${HOME?}" isn't set),
Ah at least on my windows VM "$HOME" is set in bash

Reply:
It looks like the windows doc uses powershell for the environment variable setting. The comments in `build_vk_swiftshader.sh` didn't use `$env:...`, so I'm guessing they were written for a non-windows shell?

```shell
# How it was originally written in `build_vk_swiftshader.sh`
$ set VK_ICD_FILENAMES=C:\dev\iree\build-swiftshader\Windows\vk_swiftshader_icd.json

# How it was originally written in `getting_started_windows_vulkan.md`
\```powershell
> $env:VK_ICD_FILENAMES = Resolve-Path "build-swiftshader/Windows/vk_swiftshader_icd.json"
\```
```

Could you tell me what the right way to write this in the `getting_started_windows_vulkan.md` doc would be? This is my guess for how to do it in a non-windows shell, but idk if that would work?

```
\```shell
$ set VK_ICD_FILENAMES="${USERPROFILE?}"\.swiftshader\Windows\vk_swiftshader_icd.json
\```
```

Reply:
```
λ set VK_ICD_FILENAMES=%USERPROFILE%\.swiftshader\Windows\vk_swiftshader_icd.json
λ echo %VK_ICD_FILENAMES%
C:\Users\Scott\.swiftshader\Windows\vk_swiftshader_icd.json
```

There are too many shells with differing syntax :P . I've used a few different styles depending on the audience. Powershell is generally available and nicer to work with than the base "command prompt", but it is really quirky. I use `sh.exe` and `cmd.exe` through [cmder](https://cmder.net/) almost exclusively, which with MSYS2 behaves generally like Linux. [Windows Terminal](https://github.com/microsoft/terminal) is modern and behaves pretty well in my experience.

Reply:
> It's not a dotfile though, it's a whole piece of software being installed.

hmm, other people have definitely installed entire libraries into my computer's home directory with `.` prefixes (e.g. `.bazel`, `.pyenv`, `.vscode-server`, `.oh-my-zsh`, ...). I just interpreted the `.` as a way of not being rude and adding another visible home-level directory to someone's system without their consent.

Reply:
Okay, I added this to the windows doc:
```
\```powershell
> set VK_ICD_FILENAMES=%USERPROFILE%\.swiftshader\Windows\vk_swiftshader_icd.json
\```
```